src/cascade/serialize/hdf.py

```diff
@@ -0,0 +1,286 @@
+
+class SerializationError(Exception):
+    """An error serializing data."""
+
+
+def write_var_group(hdf_group, dismod_group):
+    write_group(hdf_group, dismod_group, write_var)
+
+
+def write_grid_group(hdf_group, dismod_group):
+    write_group(hdf_group, dismod_group, write_smooth_grid)
+
+
+def write_group(hdf_group, dismod_group, writer):
+    """
+    Writes a DismodGroups object into an HDF Group. Assumes that there
+    is nothing in the HDF Group. This arranges all the names within that
+    HDF Group.
+
+    Args:
+        hdf_group (h5py.Group): The HDF Group into which to write.
+        dismod_group (DismodGroups): The Dismod objects to write.
+        writer (function): This writes whatever is in the group
+    """
+    for group_name, group in dismod_group.items():
+        for key, item in group.items():
+            if isinstance(key, str):
+                key_name = key
+                description = dict(dismod_group=group_name, key0=key, key1=None)
+            elif key[1] is None:
+                key_name = key[0]
+                description = dict(dismod_group=group_name, key0=key[0], key1=None)
+            else:
+                key_name = "_".join(str(k) for k in key)
+                description = dict(dismod_group=group_name, key0=key[0], key1=key[1])
+            name = f"{group_name}_{key_name}"
+            ds = writer(hdf_group, item, name)
+            # Add to the dataset the keys that DismodGroups uses to track
+            # each value.
+            for desc_key, desc_value in description.items():
+                if desc_value is not None:
+                    try:
+                        ds.attrs[desc_key] = desc_value
+                    except TypeError:
+                        raise RuntimeError(f"Could not write {desc_value} as attr of type {type(desc_value)}")
+
+
+def read_var_group(hdf_group, groups=None):
+    return read_group(hdf_group, read_var, groups)
+
+
+def read_grid_group(hdf_group, groups=None):
+    return read_group(hdf_group, read_smooth_grid, groups)
+
+
+def read_group(hdf_group, reader, groups=None):
+    """
+    Reads a DismodGroup of Var.
+
+    Args:
+        hdf_group (h5py.Group): The HDF Group into which to write.
+        reader (function): Reads whatever it is from the file.
+        groups (DismodGroups): A pre-existing DismodGroups.
+
+    Returns:
+        DismodGroups
+    """
+    dismod_group = groups if groups else DismodGroups()
+    for group_name, group in dismod_group.items():
+        datasets = [ds_name for ds_name in hdf_group.keys() if ds_name.startswith(group_name)]
+        for ds_name in datasets:
+            ds = hdf_group[ds_name]
+            if "dismod_group" in ds.attrs:
+                key0 = ds.attrs["key0"]
+                if "key1" in ds.attrs:
+                    try:
+                        key = (str(key0), int(ds.attrs["key1"]))
+                    except ValueError:
+                        key = (str(key0), str(ds.attrs["key1"]))
```

Comment:
This can't be typed in the underlying format?

Reply:
These are those )^@#%$#$ numpy array scalars. They are typed, but this is the only way I know to extract them.